src/rez/package_filter.py

```diff
@@ -327,7 +327,7 @@ class Rule(object):
     """Base package filter rule"""
 
     #: Rule name
-    name = None
+    name: str
```

Comment:
This pattern appears on abstract base classes in a number of places, and we have 3 choices:

## Option 1:

```python
name: str | None = None
```

typically produces many spurious mypy errors, because the code assumes the value is a string everywhere, and the `None`-case is not handled.

## Option 2:

```python
name: str
```

only safe if we're sure that the `name` attribute is always accessed on concrete sub-classes, otherwise will result in an `AttributeError` since this does not actually create an attribute. 

## Option 3:

```python
@property
@abstractmethod
def name(self) -> str:
    raise NotImplementedError
```

Provides us some guarantee that sub-classes actually implement `name`.